src/libmongoc/src/mongoc/mongoc-collection.c

```diff
@@ -3278,13 +3278,10 @@ mongoc_collection_find_and_modify_with_opts (
       }
 
       write_concern = appended_opts.writeConcern;
-   }
-
-   if (!write_concern) {
+   } else if (!write_concern) {
       if (server_stream->sd->max_wire_version >=
              WIRE_VERSION_FAM_WRITE_CONCERN &&
-          (mongoc_write_concern_is_acknowledged (collection->write_concern) ||
-           !_mongoc_client_session_in_txn (parts.assembled.session))) {
+          !_mongoc_client_session_in_txn (parts.assembled.session)) {
```

Comment:
It is safe to remove the `mongoc_write_concern_is_acknowledged (collection->write_concern)`.  The only case where 
```
server_stream->sd->max_wire_version >=
             WIRE_VERSION_FAM_WRITE_CONCERN &&
          (mongoc_write_concern_is_acknowledged (collection->write_concern) ||
           !_mongoc_client_session_in_txn (parts.assembled.session))
```
 is true and 
```
server_stream->sd->max_wire_version >=
             WIRE_VERSION_FAM_WRITE_CONCERN &&
           !_mongoc_client_session_in_txn (parts.assembled.session)
```
is false is when 
```
server_stream->sd->max_wire_version >=
             WIRE_VERSION_FAM_WRITE_CONCERN && 
mongoc_write_concern_is_acknowledged (collection->write_concern) && 
_mongoc_client_session_in_txn (parts.assembled.session)
``` 
is true.  In that case, a transaction is in progress and the call to `mongoc_cmd_parts_assemble` will always execute the check 

```
if (parts->assembled.is_txn_finish ||
          !_mongoc_client_session_in_txn (cs)) {
         _mongoc_cmd_parts_add_write_concern (parts);
 }
```
which will not add the write concern (as it should not).

Unlike we previously thought, it is not safe to remove the `!_mongoc_client_session_in_txn (parts.assembled.session)` check anywhere.  That is because the call to `mongoc_cmd_parts_assemble` will interpret that as the user attempting to send a session id with an unacknowledged write (which is forbidden) and will return an error, when actually the user is not doing that (because write concern should not propagate from the collection).

Reply:
Nicely done analysis. I agree that we cannot remove the `!_mongoc_client_session_in_txn` here or in any other places where the collection's write concern is being inherited.

> That is because the call to mongoc_cmd_parts_assemble will interpret that as the user attempting to send a session id with an unacknowledged write (which is forbidden) and will return an error, when actually the user is not doing that (because write concern should not propagate from the collection) and the function should not return an error.

Ah right, that comes from https://github.com/mongodb/specifications/blob/master/source/sessions/driver-sessions.rst#when-using-unacknowledged-writes

However, I do believe you've uncovered a bug. Though the transactions specification is [explicit about not inheriting the write concern from a client/database/collection](https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#writeconcern), the session specification does not say to bypass write concern inheritance. If a user supplies a session (that is not in a transaction) and the collection has an unacknowledged write concern, I believe an error is expected.

Further evidence: an insert_one with a session and an unacknowledged write concern does produce an error:

```
   doc = bson_new ();
   wc = mongoc_write_concern_new ();
   mongoc_write_concern_set_w (wc, MONGOC_WRITE_CONCERN_W_UNACKNOWLEDGED);
   mongoc_collection_set_write_concern (collection, wc);
   mongoc_client_session_append (client_session, opts, &error);
   r = mongoc_collection_insert_one (collection, doc, opts, NULL /* reply */, &error);
   if (!r) {
      fprintf (stderr, "Insert Error=%s\n", error.message);
   }
```

This prints `Insert Error=Cannot use client session with unacknowledged writes`. And grepping mongoc-collection.c for `collection->write_concern` indicates that would be the same behavior in all other write operations except for `mongoc_find_and_modify_with_opts()`.

Can you add a test for this? Call `mongoc_find_and_modify_with_opts()` with a session on a collection with an unacknowledged write concern, and expect an error. I believe that would fail before, but succeed with your change.

Aside: I think a lot of the hairiness of this logic points to a deeper issue within libmongoc's inheritance and validation. I think it would be simpler to have validation of options based on session and transaction state centralized in `mongoc_cmd_parts_assemble`. But, inheriting the write concern/read concern/read preference complicates that. By the time `mongoc_cmd_parts_assemble` sees those options, it does not know whether they were inherited or explicitly passed for that operation. So in cases where inheritance should not occur (e.g. a write concern should not be inherited in a transaction.

Reply:
Gah right, I misread the original condition in the if statement. I agree there was no bug originally.